Make requestStorageAccess return NoModificationAllowedError until frame is reloaded

szewai opened the pull request:
#### 65461aad5f3fbb4bd347ab2bff74ce2bdf66b2d7
<pre>
Make requestStorageAccess return NoModificationAllowedError until frame is reloaded
<a href="https://bugs.webkit.org/show_bug.cgi?id=269108">https://bugs.webkit.org/show_bug.cgi?id=269108</a>
<a href="https://rdar.apple.com/122676929">rdar://122676929</a>

Reviewed by Alex Christensen.

Add a feature flag to make requestStorageAccess return NoModificationAllowedError on granted request until frame is
reloaded, to faciliate testing behavior change on requestStorageAccess when site isolation is enabled.

Test: http/tests/storageAccess/request-throw-exception-on-grant-until-reload.html

* LayoutTests/http/tests/storageAccess/request-throw-exception-on-grant-until-reload-expected.txt: Added.
* LayoutTests/http/tests/storageAccess/request-throw-exception-on-grant-until-reload.html: Added.
* LayoutTests/http/tests/storageAccess/resources/request-throw-exception-on-grant-until-reload-iframe.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/DocumentStorageAccess.cpp:
(WebCore::DocumentStorageAccess::requestStorageAccess):
(WebCore::DocumentStorageAccess::requestStorageAccessQuirk):
* Source/WebCore/dom/DocumentStorageAccess.h:
* Source/WebCore/page/Quirks.h:
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::requestStorageAccess):
(WebKit::WebResourceLoadStatisticsStore::grantStorageAccess):
(WebKit::WebResourceLoadStatisticsStore::grantStorageAccessEphemeral):
(WebKit::WebResourceLoadStatisticsStore::grantStorageAccessInStorageSession):
(WebKit::WebResourceLoadStatisticsStore::recordFrameLoadForStorageAccess):
(WebKit::WebResourceLoadStatisticsStore::clearFrameLoadRecordsForStorageAccess):
(WebKit::WebResourceLoadStatisticsStore::storageAccessWasGrantedValueForFrame):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::scheduleResourceLoad):
(WebKit::NetworkConnectionToWebProcess::clearFrameLoadRecordsForStorageAccess):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::removeWebPageNetworkParameters):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.cpp:
(WebKit::NetworkResourceLoadParameters::NetworkResourceLoadParameters):
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::scheduleLoadFromNetworkProcess):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::removeWebFrame):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setRequestStorageAccessThrowsExceptionUntilReload):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::setRequestStorageAccessThrowsExceptionUntilReload):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):

Canonical link: <a href="https://commits.webkit.org/274636@main">https://commits.webkit.org/274636@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de87e6c2876d5b854985bfd3604d5cb000d89f76

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39599 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18578 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41955 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42134 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35499 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41905 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21468 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15907 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33064 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40173 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15662 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34270 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13585 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13563 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35224 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43411 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/33052 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35952 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35556 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39350 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/39225 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14411 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11868 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37617 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16017 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/34502 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/46231 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8881 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16065 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9427 "Found 231 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_sort3.js.default, ChakraCore.yaml/ChakraCore/test/Array/join2.js.default, ChakraCore.yaml/ChakraCore/test/ControlFlow/forInArrayAdd.js.default, ChakraCore.yaml/ChakraCore/test/Lib/uri.js.default, ChakraCore.yaml/ChakraCore/test/Object/defineProperty.js.default, ChakraCore.yaml/ChakraCore/test/Strings/property_and_index_of_string.js.default, ChakraCore.yaml/ChakraCore/test/fieldopts/fixedfieldmonocheck6.js.default, cdjs-tests.yaml/red_black_tree_test.js.dfg-eager, jsc-layout-tests.yaml/js/script-tests/dfg-arguments-osr-exit-multiple-blocks.js.layout-no-cjit, jsc-layout-tests.yaml/js/script-tests/dfg-getter-throw.js.layout-no-cjit ... (failure)") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15674 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/dom/DocumentStorageAccess.cpp:
```diff
@@ -202,7 +202,7 @@ void DocumentStorageAccess::requestStorageAccess(Ref<DeferredPromise>&& promise)
 
         // Consume the user gesture only if the user explicitly denied access.
-        bool shouldPreserveUserGesture = result.wasGranted == StorageAccessWasGranted::Yes || result.promptWasShown == StorageAccessPromptWasShown::No;
+        bool shouldPreserveUserGesture = result.wasGranted == StorageAccessWasGranted::Yes || result.wasGranted == StorageAccessWasGranted::YesWithException || result.promptWasShown == StorageAccessPromptWasShown::No;
```
We might consider making this a switch statement.  This one isn't clear, but the others are.

achristensen07 commented on Source/WebCore/dom/DocumentStorageAccess.cpp:
```diff
@@ -213,6 +213,8 @@ void DocumentStorageAccess::requestStorageAccess(Ref<DeferredPromise>&& promise)
 
         if (result.wasGranted == StorageAccessWasGranted::Yes)
             promise->resolve();
+        else if (result.wasGranted == StorageAccessWasGranted::YesWithException)
```
switch

achristensen07 commented on Source/WebCore/dom/DocumentStorageAccess.cpp:
```diff
@@ -285,7 +287,7 @@ void DocumentStorageAccess::requestStorageAccessQuirk(RegistrableDomain&& reques
         }
 
-        if (result.wasGranted == StorageAccessWasGranted::Yes)
+        if (result.wasGranted == StorageAccessWasGranted::Yes || result.wasGranted == StorageAccessWasGranted::YesWithException)
```
switch

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -525,16 +525,21 @@ void WebResourceLoadStatisticsStore::requestStorageAccessUnderOpenerEphemeral(Re
 {
     ASSERT(RunLoop::isMain());
-    postTask([this, subFrameDomain = WTFMove(subFrameDomain).isolatedCopy(), topFrameDomain = WTFMove(topFrameDomain).isolatedCopy(), frameID, pageID, promptWasShown, scope, completionHandler = WTFMove(completionHandler)]() mutable {
+    postTask([this, weakStore = ThreadSafeWeakPtr { *this }, subFrameDomain = WTFMove(subFrameDomain).isolatedCopy(), topFrameDomain = WTFMove(topFrameDomain).isolatedCopy(), frameID, pageID, promptWasShown, scope, completionHandler = WTFMove(completionHandler)]() mutable {
```
Why do we unsafely capture 'this' but also safely capture weakStore?

szewai replied:
Because postTask will hold a strong ref to this,  and postTaskReply will not (weakStore is passed to postTaskReply).

achristensen07 commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -996,9 +996,11 @@ void WebProcess::addWebFrame(FrameIdentifier frameID, WebFrame* frame)
 }
 
-void WebProcess::removeWebFrame(FrameIdentifier frameID, std::optional<WebPageProxyIdentifier> pageID)
+void WebProcess::removeWebFrame(FrameIdentifier frameID, bool isLocalFrame, std::optional<WebPageProxyIdentifier> pageID)
```
This doesn't need to be a separate parameter in this function.  Get the bool from the frame we take from m_frameMap.

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
```diff
@@ -256,8 +260,15 @@ class WebResourceLoadStatisticsStore final : public ThreadSafeRefCounted<WebReso
+    struct StorageAccessRequestRecordValue {
+        WebPageProxyIdentifier webPageProxyID;
+        Markable<WallTime> lastRequestTime;
+        WallTime lastLoadTime;
```
It seems like this should be Markable too.

szewai replied:
hmm actually this should not be Markable as the StorageAccessRequestRecordValue should be created from a frame load (i.e. the load time). Let me update the constructor.

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -1551,4 +1562,44 @@ void WebResourceLoadStatisticsStore::insertExpiredStatisticForTesting(Registrabl
+    auto& recordValue = m_storageAccessRequestRecords.ensure(key, [&]() {
+        StorageAccessRequestRecordValue value;
+        value.webPageProxyID = webPageProxyID;
+        return value;
```
This would be more future-proof if it were just return StorageAccessRequestRecordValue { webPageProxyID, std::nullopt, std::nullopt };

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
```diff
@@ -256,8 +260,15 @@ class WebResourceLoadStatisticsStore final : public ThreadSafeRefCounted<WebReso
+        WallTime lastLoadTime;
+    };
+    using StorageAccessRequestRecordKey = std::pair<WebCore::FrameIdentifier, RegistrableDomain>;
+    HashMap<StorageAccessRequestRecordKey, StorageAccessRequestRecordValue> m_storageAccessRequestRecords;
```
If we move this to NetworkConnectionToWebProcess, then there's less risk of something not cleaning up correctly and introducing an effective leak.

szewai committed:
Make requestStorageAccess return NoModificationAllowedError until frame is reloaded
https://bugs.webkit.org/show_bug.cgi?id=269108
rdar://122676929

Reviewed by Alex Christensen.

Add a feature flag to make requestStorageAccess return NoModificationAllowedError on granted request until frame is
reloaded, to faciliate testing behavior change on requestStorageAccess when site isolation is enabled.

Test: http/tests/storageAccess/request-throw-exception-on-grant-until-reload.html

* LayoutTests/http/tests/storageAccess/request-throw-exception-on-grant-until-reload-expected.txt: Added.
* LayoutTests/http/tests/storageAccess/request-throw-exception-on-grant-until-reload.html: Added.
* LayoutTests/http/tests/storageAccess/resources/request-throw-exception-on-grant-until-reload-iframe.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/DocumentStorageAccess.cpp:
(WebCore::DocumentStorageAccess::requestStorageAccess):
(WebCore::DocumentStorageAccess::requestStorageAccessQuirk):
* Source/WebCore/dom/DocumentStorageAccess.h:
* Source/WebCore/page/Quirks.h:
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::requestStorageAccess):
(WebKit::WebResourceLoadStatisticsStore::grantStorageAccess):
(WebKit::WebResourceLoadStatisticsStore::grantStorageAccessEphemeral):
(WebKit::WebResourceLoadStatisticsStore::grantStorageAccessInStorageSession):
(WebKit::WebResourceLoadStatisticsStore::recordFrameLoadForStorageAccess):
(WebKit::WebResourceLoadStatisticsStore::clearFrameLoadRecordsForStorageAccess):
(WebKit::WebResourceLoadStatisticsStore::storageAccessWasGrantedValueForFrame):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::scheduleResourceLoad):
(WebKit::NetworkConnectionToWebProcess::clearFrameLoadRecordsForStorageAccess):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::removeWebPageNetworkParameters):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.cpp:
(WebKit::NetworkResourceLoadParameters::NetworkResourceLoadParameters):
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::scheduleLoadFromNetworkProcess):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::removeWebFrame):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setRequestStorageAccessThrowsExceptionUntilReload):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::setRequestStorageAccessThrowsExceptionUntilReload):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):

Canonical link: https://commits.webkit.org/274636@main
